src/crimson/osd/osd_operation.h

```diff
@@ -379,12 +381,75 @@ class OperationThrottler : public Blocker,
   void release_throttle();
 };
 
+// the main template. by default an operation has no extenral
+// event handler (the empty tuple). specializing the template
+// allows to define backends on per-operation-type manner.
+// NOTE: basically this could be a function but C++ disallows
+// differentiating return type among specializations.
+template <class T>
+struct EventBackendRegistry {
+  template <typename...> static constexpr bool always_false = false;
+
+  static std::tuple<> get_backends() {
+    static_assert(always_false<T>, "Registry specializarion not found");
+    return {};
+  }
+};
+
+template <class T>
+constexpr static auto EVENT_NAMES = "UnknownEvent";
+
+void dump_event_default(const char name[],
```

Comment:
hmm, i am not in favor of exposing a symbol with such a general name in the public naming space... probably hide it in a dedicated internal namespace?